odl/trafos/convolution.py

```diff
@@ -0,0 +1,849 @@
+        kernel_mode, kernel_mode_in = str(kernel_mode).lower(), kernel_mode
+        if kernel_mode not in ('real', 'fourier'):
+            raise ValueError("`kernel_mode` '{}' not understood."
+                             ''.format(kernel_mode_in))
+
+        # Shift range by midpoint if the kernel is a real space element
+        if (isinstance(kernel, domain.element_type) and
+                kernel_mode == 'real'):
+            shift = kernel.space.mid_pt
+        else:
+            shift = 0
+
+        calc_range = uniform_discr_fromdiscr(domain,
+                                             min_pt=domain.min_pt + shift)
+
+        if range is None:
+            range = calc_range
+        else:
+            if range != calc_range:
+                raise ValueError('`range` {} inconsistent with range {} '
+                                 'calculated from `kernel` and `domain`'
+                                 ''.format(range, calc_range))
+
+        super().__init__(domain, range, linear=True)
+
+        # Initialize Fourier transform if not given as argument
+        self.__integral = bool(integral)
+        ft_op = kwargs.pop('ft_op', None)
+        axes = kwargs.pop('axes', None)
+        self.__ft_kwargs = kwargs.pop('ft_kwargs', {})
+        if axes is not None and 'axes' in self.__ft_kwargs:
+            raise ValueError('`axes` cannot be given both directly and in '
+                             '`ft_kwargs`')
+
+        if ft_op is None:
+            if self.integral:
+                ft_op_type = FourierTransform
+            else:
+                ft_op_type = DiscreteFourierTransform
+
+            self.__ft_op = ft_op_type(
+                self.domain, impl=ft_impl, axes=axes, **self.__ft_kwargs)
+
+        else:
+            if not isinstance(ft_op, Operator):
+                raise TypeError('`ft_op` must be an `Operator` instance, got '
+                                '{!r}'.format(ft_op))
+            if ft_op.domain != domain:
+                raise OpDomainError(
+                    '`ft_op` must have `domain` {} as its domain, got {}'
+                    ''.format(domain, ft_op.domain))
+            self.__ft_op = ft_op
+
+        # Calculate scaled kernel FT
+        kernel_kwargs = kwargs.pop('kernel_kwargs', {})
+        kernel_is_scaled = kwargs.pop('kernel_is_scaled', False)
+        scaled_ker_ft, scaling, ker_ft_op = _scaled_kernel_ft(
+            kernel, self.ft_op, kernel_mode, kernel_kwargs, kernel_is_scaled)
+
+        self.__scaled_kernel_ft = scaled_ker_ft
+        self.__kernel_scaling = scaling
+        self.__kernel_ft_op = ker_ft_op
+
+        # FT range temporary
+        ft_range_tmp = kwargs.pop('ft_range_tmp', None)
+        if ft_range_tmp is not None and ft_range_tmp not in self.ft_op.range:
+            raise OpRangeError('`ft_range_tmp` not in `ft_op.range`')
+        self.__ft_range_tmp = ft_range_tmp
+
+        # Cache adjoint operator
+        self.__adjoint = FourierSpaceConvolution(
```

Comment:
Doesnt this cause a cascade of intializations where operators initialize operators who initialize operators infinitely?